contracts/BC_fusion/StakeCredit.sol

```diff
@@ -0,0 +1,286 @@
+
+        uint256 _gasLimit = IStakeHub(STAKE_HUB_ADDR).transferGasLimit();
+        (bool success,) = delegator.call{ gas: _gasLimit, value: _totalBnbAmount }("");
+        if (!success) revert TransferFailed();
+
+        return _totalBnbAmount;
+    }
+
+    /**
+     * @dev Distribute the reward to the validator and all delegators. Only the `StakeHub` contract can call this function.
+     * @param commissionRate the commission rate of the validator
+     */
+    function distributeReward(uint64 commissionRate) external payable onlyStakeHub {
+        uint256 bnbAmount = msg.value;
+        uint256 _commission = (bnbAmount * uint256(commissionRate)) / COMMISSION_RATE_BASE;
+        uint256 _reward = bnbAmount - _commission;
+        totalPooledBNB += _reward;
+
+        // mint commission to the validator
+        _mintAndSync(validator, _commission);
+
+        emit RewardReceived(_reward, _commission);
+    }
+
+    /**
+     * @dev Slash the validator. Only the `StakeHub` contract can call this function.
+     * @param slashBnbAmount the amount of BNB to be slashed
+     * @return realSlashBnbAmount the real amount of BNB slashed
+     */
+    function slash(uint256 slashBnbAmount) external onlyStakeHub returns (uint256) {
+        uint256 selfDelegation = balanceOf(validator);
+        uint256 slashShares = getSharesByPooledBNB(slashBnbAmount);
+
+        slashShares = slashShares > selfDelegation ? selfDelegation : slashShares;
+        uint256 realSlashBnbAmount = _burnAndSync(validator, slashShares);
+
+        (bool success,) = SYSTEM_REWARD_ADDR.call{ value: realSlashBnbAmount }("");
+        if (!success) revert TransferFailed();
+
+        return realSlashBnbAmount;
+    }
+
+    /*----------------- view functions -----------------*/
+    /**
+     * @return the amount of shares that corresponds to `_bnbAmount` protocol-controlled BNB.
+     */
+    function getSharesByPooledBNB(uint256 bnbAmount) public view returns (uint256) {
+        if (totalPooledBNB == 0) revert ZeroTotalPooledBNB();
+        return (bnbAmount * totalSupply()) / totalPooledBNB;
+    }
+
+    /**
+     * @return the amount of BNB that corresponds to `_sharesAmount` token shares.
+     */
+    function getPooledBNBByShares(uint256 shares) public view returns (uint256) {
+        if (totalSupply() == 0) revert ZeroTotalShares();
+        return (shares * totalPooledBNB) / totalSupply();
+    }
+
+    /**
+     * @return the unbond request at _index and the total length of delegator's unbond queue.
+     */
+    function unbondRequest(address delegator, uint256 _index) public view returns (UnbondRequest memory, uint256) {
+        bytes32 hash = _unbondRequestsQueue[delegator].at(_index);
+        return (_unbondRequests[hash], _unbondRequestsQueue[delegator].length());
+    }
+
+    /**
+     * @return the total amount of BNB locked in the unbond queue.
+     */
+    function lockedBNBs(address delegator) public view returns (uint256) {
```

Comment:
I think it is better to have another input `number`, just like the claim function.

In case there are two many request in the queue.

Reply:
Done